Framework/MDAlgorithms/src/CompactMD.cpp

```diff
@@ -0,0 +1,140 @@
+#include "MantidMDAlgorithms/CompactMD.h"
+#include "MantidAPI/IMDIterator.h"
+using namespace Mantid::API;
+using namespace Mantid::Geometry;
+using namespace Mantid::Kernel;
+
+namespace {
+/**
+ * helper method to create a string from min and max extents (with non-zero
+ * signals) ready to be used as the PBins for IntegrateMDHistoWorkspace
+ * algorithm in
+ * exec
+ * @param minVector : Vector containing the minimum extents that we will crop
+ * to.
+ * @param maxVector : Vector containing the maximum extents that we will crop
+ * to.
+ * @param inputWs : Used in the calculation from centre to bin edges
+ * @return : a string vector of binning parameters for IntegrateMDHistoWorkspace
+ * to take as input.
+*/
+std::vector<std::string>
+createPBinStringVector(std::vector<Mantid::coord_t> minVector,
+                       std::vector<Mantid::coord_t> maxVector,
+                       IMDHistoWorkspace_sptr inputWs) {
+  size_t numDims = inputWs->getNumDims();
+  std::vector<std::string> pBinStrVector;
+  for (size_t iter = 0; iter < numDims; iter++) {
+    if (minVector[iter] >= maxVector[iter]) {
+      std::cerr << "Minimum extent of non-zero signal must be LESS than the "
+                   "maximum extent with non-zero signal" << std::endl;
+    }
+    // creating pbin string using Min and Max Centre positions
+    auto pBinStr = boost::lexical_cast<std::string>(
+                       minVector[iter] -
+                       (inputWs->getDimension(iter)->getBinWidth() * 0.5)) +
+                   ",0," +
+                   boost::lexical_cast<std::string>(
+                       maxVector[iter] +
+                       (inputWs->getDimension(iter)->getBinWidth() * 0.5));
+    pBinStrVector.push_back(pBinStr);
+  }
+  return pBinStrVector;
+}
+}
+
+namespace Mantid {
+namespace MDAlgorithms {
+
+DECLARE_ALGORITHM(CompactMD)
+
+/**
+* Finding the centre points of Bins with non-zero signal values
+* we then compare this centre to minimum and maximum centres we have
+* to get the minimum and maximum extents of the workspace that has non-zero
+* signal values in the Bins.
+* @param inputWs : The workspace that will be iterated over to find the extents.
+* @param minVec : Vector used to stored the minimum extent in each dimension
+* @param maxVec : Vector used to stored the maximum extents in each dimension
+*/
+
+void CompactMD::findFirstNonZeroMinMaxExtents(
```

Comment:
Good, but you could also have returned your min, max as a tuple boost::tuple. Even better if you typedef the templated declaration.